### ISSUE TITLE
Display and Debug Stance

### ISSUE THREAD

**Issue by trent-reed** [tpm-rs/tpm-rs#99]
In a few issues, we've talked a bit about `Display` and `Debug`, I wanted to collect our discussions on that to a single issue so that they aren't littered around everywhere (and so that we don't continue talking about those things in other issues and we can focus our discussions).

## Summary

There's some questions about code size and the opportunity to panic that have come up when talking about things like `Display` and `Debug`. We should come up with an ideal stance on the topic for tpm-rs.

## Prior Discussion

The largest contributor to this discussion has been @oddcoder, who summarized his feelings in [Issue #92](https://github.com/tpm-rs/tpm-rs/issues/92#issuecomment-2292941252):

> So I was thinking. There is an ideal solution, and the things we can do until that ideal solution is necessary.
> 
> The ideal solution in my eyes would be a fork of `ufmt`  with the following:
> 1- we have code size and code execution time benchmark.
> 2- in `std` environments `uDebug`, `uDisplay` and `uwrite!`  would still make sure that the formatting string > is the subset of the formatting string offered by `ufmt` but default to `core::fmt`.
> 3- `no-log` feature flag in the library itself where `uwrite!` would expand to no-op.
> 4- most of the `unsafe {..}` removed or isolated.
> 
> For the time being, we can just use `Debug`  and `Display` But restrict ourselves to `write!` macro, since that is the only one that can potentially be replaced by `uwrite`.
>  My justification is it wouldn't offer much sense to invest time in using most optimal logging mechanism when the real product (services/client) are not available yet.

---

Note: Discussion to follow in the rest of this issue, Only listing an introduction to the topic in this post.

**Comment by trent-reed** [comment]
## Follow-Up Questions to Investigate

1. Code size is an interesting one; when you build rust libraries, the code is trivially present, but I've usually operated under the understanding that rust is pretty good about how it cuts unused code out after linking into a binary. I'm wondering how much we can/should depend on that. We should investigate - if the unused code trimming at link time is good enough, then I don't personally see what we should worry about - imlpement what makes sense and if the client uses it, they pay the cost of using that feature (code size) - makes sense to me.
2. When talking about "execution time" and "no-log", I want to make one thing clear: I don't think that the TPM client, or the related libraries, should do *any* logging. So to me, the features like `Debug` or `Display` are simply for the client.
  - The service layer can probably log, and in that case maybe the server can have a no-log feature, but I don't want the client, base, marshalling, etc. to log - that should be handled by the caller.
  - Note that I do still want to give the callers the *ability* to log, this is more what I'm thinking of when I say "we should implement `Debug` and `Display`". This brings me back to the question of code size - if the final binary just wont have any of the display/debug logic if the user simply doesn't use it, then the feature becomes an implicit opt-in based on what the caller does (that caller could be the server (TPM), or some client (whoever is calling the TPM)). e.g. - I don't see the point in no-oping `uwrite!` for `no-log` project-wide, but maybe something like that makes sense for service layer (you can probably get by with a wrapper macro in that case, I'd think).
  - I do care to have an efficient implementation, even if the feature is opt-in, so if something like `ufmt` exists to optimize for size and avoid panics, that sounds great to me.
3. As for restricting ourselves to `write!` macro - as opposed to what? My implementations of `Display` and `Debug` have only ever used `write!`. Can you clarify what you're trying to avoid here?

**Comment by oddcoder** [comment]
> but I've usually operated under the understanding that rust is pretty good about how it cuts unused code out after linking into a binary (because it doesn't have the same side-effect promises other languages have when you link in a translation unit). I'm wondering how much we can/should depend on that.

It depends, dead code elimination, yes sure, rust will do that for you, but replacing parts of functions with more efficient (but feature incomplete) variants, they would not do that:  On example is fallibility of `fmt` and the different complex formatters for numbers and floating points, probably much more.

> but I don't want the client, base, marshalling, etc. to log

base is just bunch of types, so nothing there to log, but why not `client`?


> As for restricting ourselves to write! macro - as opposed to what?
As opposed to everything else that comes in `alloc::fmt` that is `print!`, `println!`, `format!`. ufmt does not support those and it will be problem if use them then later decide to switch

> the features like Debug or Display are simply for the caller .... I'd like to say, we should strongly try to avoid forks...I don't want to fork unless absolutely necessary.

The problem is `Debug` and `uDebug` are not compatiable, same for `Display` and `uDisplay`. If you prefer not forking `ufm`, we can derive/implement both versions. so services would find the `ufmt` variant, but code that runs on linux finds also the `core::fmt` variant.  I think this can be better than my self-proclaimed ideal solution

**Comment by oddcoder** [comment]
I must emphasize, that regardless, I believe this low priority, since we are basically planning logging for services and client at time where neither exist.

**Comment by trent-reed** [comment]
> but replacing parts of functions with more efficient (but feature incomplete) variants, they would not do that

Yeah, I don't think that is what I am talking about above, I'm strictly talking dead code elimination. If possible, I'd like for us to implement what makes sense, and leave it to the client to use what they actually want/need.

> base is just bunch of types, so nothing there to log, but why not client?

I can't see a good reason to log in client, I'd rather return errors that are good enough for the caller to know what went wrong. I think this is more doable than it sounds, and won't be that bad at all. Also, I don't think it's a good idea for `client` code to log as a default-stance, because that's a bit presumptuous of what the caller may or may not want.

I highly suspect that we don't need logging in `client`, I'd like to start from that, then we'll talk about it if we actually do find a need for it.

> I must emphasize, that regardless, I believe this low priority, since we are basically planning logging for services and client at time where neither exist.

Yea, I just want to know what we can do *now*. So if we're okay with it, I'd like for the current outcome of this conversation to be:

*   Let's just freely implement `Debug` and `Display`, and tackle it more appropriately if needed later.

The reason it's kind of important to talk about now is for testing. Right now I'm seeing some test results on failures on code change that are super unhelpful, so I wanted to see how to make that better.

Thanks!

**Comment by oddcoder** [comment]
Hello,
Apologies for the delay, I was in some conference. In retrospect I should have informed you earlier. 

> Let's just freely implement Debug and Display, and tackle it more appropriately if needed later.

You are right, We should not keep this waiting any longer with no reasonable justification,  feel free to implement.
